File: simple_python_demo.py
#!/usr/bin/env python3
"""
Simple Python syntax demonstration for compiler testing
"""

def add_numbers(a, b):
    """Add two numbers"""
    return a + b

def factorial(n):
    """Calculate factorial"""
    if n <= 1:
        return 1
    else:
        return n * factorial(n - 1)

class Calculator:
    """Simple calculator class"""
    
    def __init__(self, name):
        self.name = name
        self.history = []
    
    def add(self, a, b):
        result = a + b
        self.history.append(f"add({a}, {b}) = {result}")
        return result
    
    def multiply(self, a, b):
        result = a * b
        self.history.append(f"multiply({a}, {b}) = {result}")
        return result
    
    def get_history(self):
        return self.history

def fibonacci_generator(n):
    """Generate fibonacci sequence"""
    a, b = 0, 1
    for i in range(n):
        yield a
        a, b = b, a + b

def list_comprehension_demo():
    """Demonstrate list comprehensions"""
    # Simple list comprehension
    squares = [x**2 for x in range(10)]
    
    # List comprehension with condition
    even_squares = [x**2 for x in range(10) if x % 2 == 0]
    
    # Nested list comprehension
    matrix = [[i*j for j in range(1, 4)] for i in range(1, 4)]
    
    return squares, even_squares, matrix

def dict_comprehension_demo():
    """Demonstrate dictionary comprehensions"""
    # Simple dict comprehension
    square_dict = {x: x**2 for x in range(5)}
    
    return square_dict

def exception_handling_demo(a, b):
    """Demonstrate exception handling"""
    try:
        result = a / b
        print(f"Division successful: {a} / {b} = {result}")
        return result
    except ZeroDivisionError:
        print("Error: Cannot divide by zero")
        return None
    except Exception as e:
        print(f"Error: {e}")
        return None

def main():
    """Main function"""
    print("=== Python Syntax Demo ===")
    
    # Basic function calls
    print(f"add_numbers(5, 3) = {add_numbers(5, 3)}")
    print(f"factorial(5) = {factorial(5)}")
    
    # Class usage
    calc = Calculator("MyCalc")
    print(f"calc.add(10, 5) = {calc.add(10, 5)}")
    print(f"calc.multiply(4, 7) = {calc.multiply(4, 7)}")
    print(f"Calculator history: {calc.get_history()}")
    
    # Generator
    fib_list = list(fibonacci_generator(10))
    print(f"Fibonacci(10): {fib_list}")
    
    # List comprehensions
    squares, even_squares, matrix = list_comprehension_demo()
    print(f"Squares: {squares}")
    print(f"Even squares: {even_squares}")
    print(f"Matrix: {matrix}")
    
    # Dictionary comprehension
    square_dict = dict_comprehension_demo()
    print(f"Square dict: {square_dict}")
    
    # Exception handling
    print("Exception handling:")
    exception_handling_demo(10, 2)
    exception_handling_demo(10, 0)
    
    print("\n=== Demo completed ===")

if __name__ == "__main__":
    main()

File: simple_python_demo.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


class Calculator {
    public:
        Calculator(std::string name) {
            this->name = name;
            this->history = std::vector<int>{};
        }
        int add(std::string a, std::string b) {
    auto result = a + b;
    this->history.push_back(f"add({a}, {b}) = {result}");
    return result;
}


        int multiply(std::string a, std::string b) {
    auto result = a * b;
    this->history.push_back(f"multiply({a}, {b}) = {result}");
    return result;
}


        int get_history() {
    return this->history;
}


    public:
        std::string name = "";
        std::vector<int> history;
        std::string history.append(f"add({a}, {b}) = "";
        std::string history.append(f"multiply({a}, {b}) = "";
};


int add_numbers(int a, int b) {
    /* docstring */
    return a + b;
}


int factorial(int n) {
    /* docstring */
    if (n <= 1) {
        return 1;
    }
    else {
        return n * factorial(n - 1);
    }
}


void fibonacci_generator(int n) {
    /* docstring */
    auto a, b = 0, 1;
    for (int i = 0; i < n; ++i) {
        /* yield */
        auto a, b = b, a + b;
    }
}


int list_comprehension_demo() {
    /* docstring */
    auto squares = 0;
    auto even_squares = 0;
    auto matrix = 0;
    return squares, even_squares, matrix;
}


int dict_comprehension_demo() {
    /* docstring */
    auto square_dict = std::unordered_map<std::string, int>{{"x", 0}};
    return square_dict;
}


int exception_handling_demo(std::string a, std::string b) {
    bool __fluxus_exc=false;
    /* docstring */
    {
        auto result = 0;
if ((b) == 0) { __fluxus_exc = true; result = 0; }
else { result = ((1.0*(a))/(b)); }

        std::cout << to_str("Division successful: ") << to_str(a) << to_str(" / ") << to_str(b) << to_str(" = ") << to_str(result) << std::endl;
        return result;
    }
    if (__fluxus_exc) {
        std::cout << "Error: Cannot divide by zero" << std::endl;
        return 0;
    }
    if (__fluxus_exc) { auto e = 0;
        std::cout << to_str("Error: ") << to_str(e) << std::endl;
        return 0;
    }
}


int main() {
    /* docstring */
    std::cout << "=== Python Syntax Demo ===" << std::endl;
    std::cout << to_str("add_numbers(5, 3) = ") << to_str(add_numbers(5, 3)) << std::endl;
    std::cout << to_str("factorial(5) = ") << to_str(factorial(5)) << std::endl;
    auto calc = Calculator("MyCalc");
    std::cout << to_str("calc.add(10, 5) = ") << to_str(calc.add(10, 5)) << std::endl;
    std::cout << to_str("calc.multiply(4, 7) = ") << to_str(calc.multiply(4, 7)) << std::endl;
    std::cout << to_str("Calculator history: ") << to_str(calc.get_history()) << std::endl;
    auto fib_list = list(fibonacci_generator(10));
    std::cout << to_str("Fibonacci(10): ") << to_str(fib_list) << std::endl;
    auto squares, even_squares, matrix = list_comprehension_demo();
    std::cout << to_str("Squares: ") << to_str(squares) << std::endl;
    std::cout << to_str("Even squares: ") << to_str(even_squares) << std::endl;
    std::cout << to_str("Matrix: ") << to_str(matrix) << std::endl;
    auto square_dict = dict_comprehension_demo();
    std::cout << to_str("Square dict: ") << to_str(square_dict) << std::endl;
    std::cout << "Exception handling:" << std::endl;
    exception_handling_demo(10, 2);
    exception_handling_demo(10, 0);
    std::cout << "\n=== Demo completed ===" << std::endl;
    return 0;
}


